packages/next/src/server/app-render/app-render.tsx

```diff
@@ -987,7 +987,12 @@ async function renderToHTMLOrFlightImpl(
         }
 
         return { stream }
-      } catch (err) {
+      } catch (caughtError: any) {
+        const originalError = capturedErrors.find(
+          (e: any) => e.digest === caughtError.digest
+        )
```

Comment:
was talking to Wyatt about this today. We probably want to make this into a map so we can just look up by digest.

We also should make sure that our digests are unique. At least for anything that goes into allCapturedErrors. This may already be the case b/c the redirect and notFound errors may be omitted (I haven't looked closely yet). However there is a case where I think we could still get the same digest for multiple onError calls. When you abort a react render all the suspense boundaries still pending are marked for client render and each one results in an onError call. But since the error object will be the same in each case it will produce an identical digest. We actually would have a component stack in dev and when prerendering which can make them unique again but it won't help with regular renders in prod. Perhaps we just add a bit of entropy to the digest on each onError calls since we know each call to onError represents a unique error even if the error value is shared.

Reply:
Continuing on the reasons why a map might be better in this case...

I'd expect that if an error occurs, because the error digest is unique, we should not insert into the map if the digest already exists. This is really important, as different renderers may be handed the "masked" error rather than the underlying error first. For example:

1. Original `RedirectError` is thrown, containing a digest of `NEXT_REDIRECT;replace;/some/url;303;`
2. Add error to error map with the digest as the key
3. The `onError` returns this digest, which is later consumed by other layers, which in turn is caught by it's `onError` handler

In this case, the first occurrence of the error should be the one that is the original error, not any subsequent ones. _Technically_, your code here works, but that's only because _new_ errors are pushed into the array, and the `.find()` will get the _first_ occurrence of that digest.